keps/sig-scheduling/5027-dra-admin-controlled-device-attributes/README.md

```diff
@@ -0,0 +1,857 @@
+can only decide for themselves whether they ignore taints and do so at their
+own risk. They cannot break other users' workloads.
+
+Performance in the scheduler could be an issue. This will be mitigated by
+(re-)applying overrides only when they or the device definitions change, which
+should be rare.
+
+## Design Details
+
+### API
+
+The ResourceSlicePatch is a cluster-scoped type in the `resource.k8s.io` API
+group, initially in `v1alpha3` (the alpha version in Kubernetes 1.32). Because
+it may be useful to clean up after disabling the feature and because the
+device taint feature also uses this type, it gets served unconditionally as long as
+the `v1alpha3` version is enabled. Fields related specifically to this KEP
+are feature-gated.
+
+```Go
+type ResourceSlicePatch struct {
+metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    // Changing the spec automatically increments the metadata.generation number.
+    Spec ResourceSlicePatchSpec
+}
+
+type ResourceSlicePatchSpec struct {
+    // Devices defines how to override device attributes and taints.
+    Devices DevicePatch
+}
+
+// DevicePatch selects one or more devices by class, driver, pool, device names
+// and/or CEL selectors. All of these criteria must be satisfied by a device, otherwise
+// it is ignored by the override. A DevicePatch with no selection criteria is
+// valid and matches all devices.
+type DevicePatch struct {
+    // Selector defines which device(s) the overrides are applied to.
+    //
+    // +optional
+    Selector *DevicePatchSelector
+
+    // If a ResourceSlice and a DevicePatch define the same attribute or
+    // capacity, the value of the DevicePatch is used. If multiple
+    // different DevicePatchs match the same device, then the one with
+    // the highest priority wins. If the priorities are the same, it is non-deterministic
+    // which override is used.
+    Priority int
+
+    // Attributes defines the set of attributes to override for matching devices.
+    // The name of each attribute must be unique in that set and
+    // include the domain prefix.
+    //
+    // The maximum number of attributes and capacities combined is 32.
+    // This is an alpha field and requires enabling the DRAAdminControlledDeviceAttributes
+    // feature gate.
+    //
+    // +optional
+    // +featureGate:DRAAdminControlledDeviceAttributes
+    Attributes map[FullyQualifiedName]DeviceAttribute
+
+    // Capacity defines the set of capacities to override for matching devices.
+    // The name of each capacity must be unique in that set and
+    // include the domain prefix.
+    //
+    // The maximum number of attributes and capacities combined is 32.
+    // This is an alpha field and requires enabling the DRAAdminControlledDeviceAttributes
+    // feature gate.
+    //
+    // +optional
+    // +featureGate:DRAAdminControlledDeviceAttributes
+    Capacity map[FullyQualifiedName]DeviceCapacity
```

Comment:
Is the intent here to match the `capacity` field for devices in a ResourceSlice? 
https://github.com/kubernetes/kubernetes/blob/45d0fddaf1f24f7b559eb936308ce2aeb9871850/staging/src/k8s.io/api/resource/v1alpha3/types.go#L225
```suggestion
    Capacity map[FullyQualifiedName]resource.Quantity
```

Reply:
nvm I see v1beta1 has a distinct `DeviceCapacity` type. https://github.com/kubernetes/kubernetes/blob/45d0fddaf1f24f7b559eb936308ce2aeb9871850/staging/src/k8s.io/api/resource/v1beta1/types.go#L224-L236